### PR TITLE
use Variadic templates

### PR DESCRIPTION
**Pull request by sloriot**
@mglisse: Not sure why you have two versions in `Kernel_d/include/CGAL/Kernel_d/function_objects.h`, one with different template parameters and one with identical one.

### CODE REVIEW

**Comment by mglisse** [comment]
> @mglisse: Not sure why you have two versions in `Kernel_d/include/CGAL/Kernel_d/function_objects.h`, one with different template parameters and one with identical one.

I wrote NewKernel_d, I didn't have anything to do with Kernel_d...
Some guesses: maybe it was getting too tiring to have different params? maybe with 6+ params it is always something like Point(coor1, coor2...) whereas with fewer it may be a mix of iterators, numbers, etc?
Anyway, making it generic seems right, although I would go one step further and use "perfect" forwarding.

File: Filtered_kernel/include/CGAL/Kernel_profiler.h [hunk 1]
```diff
@@ -43,77 +43,13 @@ struct Primitive_profiler
...
 
-    template <class A1>
+    template <class ... A>
     result_type
-    operator()(const A1 &a1) const
+    operator()(const A&& ... a) const
     {
 	CGAL_KERNEL_PROFILER;
-	return P::operator()(a1);
+	return P::operator()(std::forward(a)...);
```
Comment (lrineau) [inline Filtered_kernel/include/CGAL/Kernel_profiler.h]: Fix the perfect forwarding:
```suggestion
	return P::operator()(std::forward<A>(a)...);
```

File: Kernel_d/include/CGAL/Kernel_d/function_objects.h [hunk 2]
```diff
@@ -41,81 +41,10 @@ class Construct
...
-    ToBeConstructed
-    operator()( const A& a1, const A& a2, const A& a3,
-                const A& a4, const A& a5, const A& a6,
-                const A& a7, const A& a8, const A& a9,
-                const A& a10,const A& a11,const A& a12,
-                const A& a13) const
-    { return ToBeConstructed(a1,a2,a3,a4,a5,a6,a7,a8,a9,a10,a11,a12,a13); }
-
+    operator()( const A&& ... a) const
+    { return ToBeConstructed(std::forward(a)...); }
```
Comment (lrineau) [inline Kernel_d/include/CGAL/Kernel_d/function_objects.h]:
```suggestion
    { return ToBeConstructed(std::forward<A>(a)...); }
```

File: Straight_skeleton_2/include/CGAL/Straight_skeleton_2/Straight_skeleton_builder_traits_2_aux.h [hunk 3]
```diff
@@ -115,98 +115,23 @@ class Exceptionless_filtered_construction
...
-      FC_result_type fr = Filter_construction(To_Filtered(a1));
+      FC_result_type fr = Filter_construction(To_Filtered(std::forward(a))...);
       if ( fr )
         return From_Filtered(fr);
     }
     catch (Uncertain_conversion_exception&) {}
 
     Protect_FPU_rounding<!Protection> P(CGAL_FE_TONEAREST);
-    EC_result_type er = Exact_construction(To_Exact(a1)) ;
+    EC_result_type er = Exact_construction(To_Exact(std::forward(a))...) ;
```
Comment (lrineau) [inline Straight_skeleton_2/include/CGAL/Straight_skeleton_2/Straight_skeleton_builder_traits_2_aux.h]:
```suggestion
    EC_result_type er = Exact_construction(To_Exact(std::forward<A>(a))...) ;
```

File: Straight_skeleton_2/include/CGAL/Unfiltered_predicate_adaptor.h [hunk 4]
```diff
@@ -54,111 +54,14 @@ class Unfiltered_predicate_adaptor
...
-  result_type
-  operator()(const A1 &a1, const A2 &a2, const A3 &a3, const A4 &a4,
-             const A5 &a5, const A6 &a6, const A7 &a7, const A8 &a8,
-             const A9 &a9 ) const
-#ifndef CGAL_CFG_OUTOFLINE_TEMPLATE_MEMBER_DEFINITION_BUG
-  ;
-#else
-  {
-    return static_cast<result_type>(Certified_approx_predicate(a1, a2, a3, a4, a5, a6, a7, a8, a9));
+    return static_cast<result_type>(Certified_approx_predicate(std::forward(a)...));
```
Comment (lrineau) [inline Straight_skeleton_2/include/CGAL/Unfiltered_predicate_adaptor.h]:
```suggestion
    return static_cast<result_type>(Certified_approx_predicate(std::forward<A>(a)...));
```

File: Straight_skeleton_2/include/CGAL/Unfiltered_predicate_adaptor.h [hunk 5]
```diff
@@ -167,90 +70,12 @@ class Unfiltered_predicate_adaptor
...
+  template <class ... A>
 typename Unfiltered_predicate_adaptor<CAP>::result_type
 Unfiltered_predicate_adaptor<CAP>::
-  operator()(const A1 &a1, const A2 &a2, const A3 &a3, const A4 &a4,
-             const A5 &a5, const A6 &a6, const A7 &a7, const A8 &a8, 
-             const A9 &a9 ) const
+  operator()(const A&& ... a) const
 {
-  return static_cast<result_type>(Certified_approx_predicate(a1, a2, a3, a4, a5, a6, a7, a8, a9));
+  return static_cast<result_type>(Certified_approx_predicate(std::forward(a)...));
```
Comment (lrineau) [inline Straight_skeleton_2/include/CGAL/Unfiltered_predicate_adaptor.h]:
```suggestion
  return static_cast<result_type>(Certified_approx_predicate(std::forward<A>(a)...));
```

File: Filtered_kernel/include/CGAL/Kernel_profiler.h [hunk 6]
```diff
@@ -43,77 +43,13 @@ struct Primitive_profiler
     Primitive_profiler(const P& p = P())
       : P(p) {}
 
-    template <class A1>
+    template <class ... A>
     result_type
-    operator()(const A1 &a1) const
+    operator()(const A&& ... a) const
```
Comment (theartful) [inline Filtered_kernel/include/CGAL/Kernel_profiler.h]: here

File: Kernel_d/include/CGAL/Kernel_d/function_objects.h [hunk 7]
```diff
@@ -41,81 +41,10 @@ class Construct
...
-    template <class A> 
-    ToBeConstructed
-    operator()( const A& a1, const A& a2, const A& a3,
-                const A& a4, const A& a5, const A& a6,
-                const A& a7, const A& a8, const A& a9,
-                const A& a10,const A& a11,const A& a12,
-                const A& a13) const
-    { return ToBeConstructed(a1,a2,a3,a4,a5,a6,a7,a8,a9,a10,a11,a12,a13); }
-
+    operator()( const A&& ... a) const
```
Comment (theartful) [inline Kernel_d/include/CGAL/Kernel_d/function_objects.h]: here

File: Straight_skeleton_2/include/CGAL/Straight_skeleton_2/Straight_skeleton_builder_traits_2_aux.h [hunk 8]
```diff
@@ -115,98 +115,23 @@ class Exceptionless_filtered_construction
 
   Exceptionless_filtered_construction() {}
 
-  template <class A1>
+  template <class ... A>
   result_type
-  operator()(const A1 &a1) const
+  operator()(const A&& ... a) const
```
Comment (theartful) [inline Straight_skeleton_2/include/CGAL/Straight_skeleton_2/Straight_skeleton_builder_traits_2_aux.h]: here

File: Straight_skeleton_2/include/CGAL/Unfiltered_predicate_adaptor.h [hunk 9]
```diff
@@ -54,111 +54,14 @@ class Unfiltered_predicate_adaptor
     : Certified_approx_predicate(o1, o2)
   {}
 
-  template <class A1>
+  template <class ... A>
   result_type
-  operator()(const A1 &a1) const
+  operator()(const A&& ... a) const
```
Comment (theartful) [inline Straight_skeleton_2/include/CGAL/Unfiltered_predicate_adaptor.h]: here

File: Straight_skeleton_2/include/CGAL/Unfiltered_predicate_adaptor.h [hunk 10]
```diff
@@ -167,90 +70,12 @@ class Unfiltered_predicate_adaptor
...
-
-template <class CAP>
-  template <class A1, class A2, class A3, class A4, class A5, class A6, class A7, class A8, class A9>
+  template <class ... A>
 typename Unfiltered_predicate_adaptor<CAP>::result_type
 Unfiltered_predicate_adaptor<CAP>::
-  operator()(const A1 &a1, const A2 &a2, const A3 &a3, const A4 &a4,
-             const A5 &a5, const A6 &a6, const A7 &a7, const A8 &a8, 
-             const A9 &a9 ) const
+  operator()(const A&& ... a) const
```
Comment (theartful) [inline Straight_skeleton_2/include/CGAL/Unfiltered_predicate_adaptor.h]: and here

**Comment by sloriot** [comment]
thanks @theartful

**Comment by maxGimeno** [comment]
Successfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-5.1-Ic-100.shtml

**Pull request merged**